paasta_tools/iptables.py

```diff
@@ -0,0 +1,113 @@
+"""iptables helper functions.
+
+Unlike the `firewall` module, these functions know nothing about PaaSTA and
+could effectively be a third-party library. They just make working with
+iptables a little bit easier.
+"""
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import re
+import shlex
+import subprocess
+
+
+class ChainDoesNotExist(Exception):
+    pass
+
+
+def all_chains():
+    output = subprocess.check_output(
+        ('iptables', '-t', 'filter', '--list-rules'),
+    )
+    chain_regex = re.compile('-N ([^ ]+)$')
+    chains = set()
+    for line in output.splitlines():
+        m = chain_regex.match(line)
+        if m:
+            chains.add(m.group(1))
+    return chains
+
+
+def ensure_chain(chain, rules=None):
+    """Idempotently ensure a chain exists and has an exact set of rules.
+
+    This function creates or updates an existing chain to match the rules
+    passed in.
+
+    This function will not reorder existing rules, but any new rules are always
+    inserted at the front of the chain.
+    """
+    try:
+        current_rules = list_chain(chain)
+    except ChainDoesNotExist:
+        create_chain(chain)
+        current_rules = set()
+
+    if rules is not None:
+        for rule in rules:
+            if rule not in current_rules:
+                add_rule(chain, rule)
+
+        for rule in current_rules:
+            if rule not in rules:
```

Comment:
Yeah, that mental model is right. Just to be explicit, the reason for that is that when inserting rules, the insertion order actually matters (the REJECT needs to be inserted first). Turning it into a set during the second stage sounds like a good idea, I'll do that.

iptables definitely normalizes the rules when they are printed. That's also why I've written some of the rules in a slightly strange way. For example if you define a rule `-A INPUT -p tcp --dport 8888  -j REJECT`, it actually creates a rule that says `-A INPUT -p tcp -m tcp --dport 8888 -j REJECT --reject-with icmp-port-unreachable`.

My main concern would be if iptables changes the normalization, which could conceivably happen during e.g. an OS upgrade. In that case, this code would see the right rules but with the wrong normalization and remove/re-add them on every run. I don't have a great solution for that besides parsing the rules. Does that seem like a good idea to you?